tests/unit/completion/test_completer.py

```diff
@@ -0,0 +1,136 @@
+        super().__init__(parent)
+        self.kind = kind
+
+
+@pytest.fixture
+def cmd(stubs):
+    """Create the statusbar command prompt the completer uses."""
+    return stubs.FakeStatusbarCommand()
+
+
+@pytest.fixture
+def completer(qtbot, cmd, config_stub):
+    """Create the completer used for testing."""
+    config_stub.data = {'completion': {'auto-open': False}}
+    return Completer(cmd, 0)
+
+
+@pytest.fixture
+def instances(monkeypatch):
+    """Mock the instances module so get returns a fake completion model."""
+    # populate a model for each completion type, with a nested structure for
+    # option and value completion
+    instances = {kind: FakeCompletionModel(kind) for kind in Completion}
+    instances[Completion.option] = {
+        'general': FakeCompletionModel(Completion.option),
+    }
+    instances[Completion.value] = {
+        'general': {
+            'ignore-case': FakeCompletionModel(Completion.value),
+        }
+    }
+    monkeypatch.setattr('qutebrowser.completion.completer.instances',
+                        instances)
+    return instances
+
+
+@pytest.fixture
+def cmdutils_patch(monkeypatch, stubs):
+    """Patch the cmdutils module to provide fake commands."""
+    cmds = {
+        'set': [Completion.section, Completion.option, Completion.value],
+        'help': [Completion.helptopic],
+        'quickmark-load': [Completion.quickmark_by_name],
+        'bookmark-load': [Completion.bookmark_by_url],
+        'open': [Completion.url],
+        'buffer': [Completion.tab],
+        'session-load': [Completion.sessions],
+        'bind': [Completion.empty, Completion.command],
+    }
+    cmd_utils = stubs.FakeCmdUtils({
+        name: stubs.FakeCommand(completion=compl)
+        for name, compl in cmds.items()
+    })
+    monkeypatch.setattr('qutebrowser.completion.completer.cmdutils',
+                        cmd_utils)
+    return cmd_utils
+
+
+@pytest.mark.parametrize('txt, expected', [
+    (':nope|', Completion.command),
+    (':nope |', None),
+    (':set |', Completion.section),
+    (':set gen|', Completion.section),
+    (':set general |', Completion.option),
+    (':set what |', None),
+    (':set general ignore-case |', Completion.value),
+    (':set general huh |', None),
+    (':help |', Completion.helptopic),
+    (':quickmark-load |', Completion.quickmark_by_name),
+    (':bookmark-load |', Completion.bookmark_by_url),
+    (':open |', Completion.url),
+    (':buffer |', Completion.tab),
+    (':session-load |', Completion.sessions),
+    (':bind |', Completion.empty),
+    (':bind <c-x> |', Completion.command),
+    (':bind <c-x> foo|', Completion.command),
+    (':bind <c-x>| foo', Completion.empty),
+    (':set| general ', Completion.command),
+    (':|set general ', Completion.command),
+    (':set gene|ral ignore-case', Completion.section),
+    (':|', Completion.command),
+    (':   |', Completion.command),
+    (':bookmark-load      |', Completion.bookmark_by_url),
```

Comment:
Some other tests which come to mind:

- Commandline starting with `/` (i.e. a search)
- With a `--flag` (like `--mode` with bind) and cursor after the flag
- Same, with cursor on the flag
- Only a flag (`:--foo|` or so)
- Command which doesn't have a completion
- Something like `:foo | bar`

(Some of them are inspired by the [coverage report](https://codecov.io/gh/The-Compiler/qutebrowser/src/07edcce697a489619a117cfdd02e1ef1ff2636d5/qutebrowser/completion/completer.py) for `update_cursor_part`, `update_completion`, `_filter_cmdline_parts` and `_get_new_completion` - you might be able to find more I didn't think of)